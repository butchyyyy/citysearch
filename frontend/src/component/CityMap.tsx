import CityMarker from "component/CityMarker"
import City from "model/City"
import React from "react"
import { GoogleMap, withGoogleMap, WithGoogleMapProps, withScriptjs, WithScriptjsProps } from "react-google-maps"
import { MarkerWithLabel } from "react-google-maps/lib/components/addons/MarkerWithLabel"

interface OwnProps {
  cities: City[]
}

interface WrapperProps {
  mapApiKey: string
}

type Props = OwnProps & WithGoogleMapProps & WithScriptjsProps

class CityMap extends React.PureComponent<Props> {
  mapRef: React.RefObject<GoogleMap>

  constructor(props) {
    super(props)
    this.mapRef = React.createRef()
  }

  render() {
    return (
        <GoogleMap
            defaultCenter={{ lat: 0, lng: 0 }}
            defaultZoom={2}
            ref={this.mapRef}
        >
          {this.props.cities.map((city, index) => (
              <CityMarker key={index} city={city} />
          ))}
        </GoogleMap>
    )
  }

  componentDidUpdate() {
    if (this.props.cities && this.props.cities.length > 0) {
      const bounds = new google.maps.LatLngBounds()
      this.props.cities.forEach((city) => {
        bounds.extend({ lat: city.lat, lng: city.lng })
      })
      this.mapRef.current.fitBounds(bounds)
    }
  }
}

const Enhanced = withScriptjs(withGoogleMap(CityMap))

const composed = (props: OwnProps & WrapperProps) => (
    <Enhanced
        cities={props.cities}
        googleMapURL={`https://maps.googleapis.com/maps/api/js?key=${props.mapApiKey}`}
        loadingElement={<div style={{ height: `100%` }} />}
        containerElement={<div style={{ height: `600px` }} />}
        mapElement={<div style={{ height: `100%` }} />}
    />
)

export default composed
