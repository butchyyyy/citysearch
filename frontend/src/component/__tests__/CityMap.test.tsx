import { testCities } from "__tests__/TestData"
import CityMap from "component/CityMap"
import { shallow } from "enzyme"
import React from "react"

describe("Test CityMap Component", () => {
  it("Test CityMap Snapshot", () => {
    const wrapper = shallow(<CityMap cities={testCities} mapApiKey="" markerClusterer />)
    expect(wrapper).toMatchSnapshot()
  })
})
