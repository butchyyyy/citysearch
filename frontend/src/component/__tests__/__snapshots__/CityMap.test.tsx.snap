// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CityMap Component Test CityMap Snapshot 1`] = `
<withScriptjs(withGoogleMap(CityMap))
  cities={
    Array [
      Object {
        "country": "Czechia",
        "lat": 50.08804,
        "lng": 14.42076,
        "name": "Prague",
        "population": 1165581,
      },
      Object {
        "country": "United States, Minnesota, Minneapolis",
        "lat": 44.5433,
        "lng": -93.57607,
        "name": "New Prague",
        "population": 7582,
      },
      Object {
        "country": "Some Country",
        "lat": 0,
        "lng": 0,
        "name": "I don't have a population",
        "population": null,
      },
    ]
  }
  containerElement={
    <div
      style={
        Object {
          "height": "600px",
        }
      }
    />
  }
  googleMapURL="https://maps.googleapis.com/maps/api/js?key="
  loadingElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  mapElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  markerClusterer={true}
/>
`;
